### PR TITLE
build: track generated CSV, and validate

### PR DESCRIPTION
**Pull request by BlaineEXE**
Start tracking the generated CSV, and validate that they don't change in
builds.

Signed-off-by: Blaine Gardner <blaine.gardner@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: images/ceph/Makefile [hunk 1]
```diff
@@ -66,12 +66,6 @@ do.build:
 	@cp $(MANIFESTS_DIR)/test-data/ceph-status-out $(TEMP)/rook-external/test-data/
 	@$(MAKE) list-image
 
-ifeq ($(INCLUDE_CSV_TEMPLATES),true)
-	@$(MAKE) CSV_TEMPLATE_DIR=$(TEMP) generate-csv-templates
-	@cp -r $(TEMP)/cluster/olm/ceph/templates $(TEMP)/ceph-csv-templates
-else
-	mkdir $(TEMP)/ceph-csv-templates
-endif
```
Comment (BlaineEXE) [inline images/ceph/Makefile:74]: Now that we can generate CSVs separately, let's separate this. When the build also does this and changes files in the same location, it causes conflicts when checking for updates

File: Makefile [hunk 2]
```diff
@@ -115,7 +115,7 @@ do.build.platform.%:
 
 do.build.parallel: $(foreach p,$(PLATFORMS), do.build.platform.$(p))
 
-build: csv-clean build.common ## Only build for linux platform
```
Comment (BlaineEXE) [inline Makefile:118]: csv-clean here was causing changes to tracked CSVs on build.

File: Makefile [hunk 3]
```diff
@@ -168,14 +168,29 @@ prune: ## Prune cached artifacts.
 csv-ceph: export MAX_DESC_LEN=0 # sets the description length to 0 since CSV cannot be bigger than 1MB
 csv-ceph: export NO_OB_OBC_VOL_GEN=true
 csv-ceph: csv-clean crds ## Generate a CSV file for OLM.
+	$(MAKE) -C images/ceph csv-clean
 	$(MAKE) -C images/ceph csv
 
-csv-clean: ## Remove existing OLM files.
-	@$(MAKE) -C images/ceph csv-clean
```
Comment (BlaineEXE) [inline Makefile:174]: csv-clean in general was causing changes to tracked CSVs on build, so I just remove it.

File: Makefile [hunk 4]
```diff
@@ -168,14 +168,29 @@ prune: ## Prune cached artifacts.
...
-
+GEN_CRD_TEMP := /tmp/rook-ceph-gen-crds
+BUILD_CRDS_INTO_DIR ?= $(GEN_CRD_TEMP) # unless overridden, build CRDs into the temp dir
```
Comment (BlaineEXE) [inline Makefile:175]: So the intermediate steps don't interfere with tracked CSVs.

File: images/ceph/Makefile [hunk 5]
```diff
@@ -89,9 +89,10 @@ generate-csv-templates: $(OPERATOR_SDK) $(YQ) ## Generate CSV templates for OLM
 	@# first, copy the existing CRDs and OLM catalog directory to CSV_TEMPLATE_DIR
 	@# then, generate or copy all prerequisites into CSV_TEMPLATE_DIR (e.g., CRDs)
 	@# finally, generate the templates in-place using CSV_TEMPLATE_DIR as a staging dir
-	@mkdir -p $(CSV_TEMPLATE_DIR)
-	@cp -a ../../cluster $(CSV_TEMPLATE_DIR)/cluster
-	@set -eE;\
+	@ echo yq=$(YQ) operator-sdk=$(OPERATOR_SDK)
+	mkdir -p $(CSV_TEMPLATE_DIR)
+	cp -a ../../cluster $(CSV_TEMPLATE_DIR)/cluster
+	set -eE;\
```
Comment (BlaineEXE) [inline images/ceph/Makefile:95]: Removing `@` symbols here and elsewhere was useful in debugging build issues in the CI for this PR and will be useful in the future as well, so I keep them.

File: cluster/olm/ceph/deploy/operator.yaml [hunk 6]
```diff
@@ -0,0 +1,151 @@
+# OLM: BEGIN OPERATOR DEPLOYMENT
```
Comment (leseb) [inline cluster/olm/ceph/deploy/operator.yaml:1]: Why is the operator's configmap gone?

File: cluster/olm/ceph/generate-rook-csv.sh [hunk 7]
```diff
@@ -124,13 +127,19 @@ function generate_csv(){
     "${YQ_CMD_DELETE[@]}" "$CSV_FILE_NAME" 'spec.maintainers[0]'
 
     "${YQ_CMD_MERGE_OVERWRITE[@]}" "$CSV_FILE_NAME" "$ASSEMBLE_FILE_COMMON"
-    "${YQ_CMD_WRITE[@]}" "$CSV_FILE_NAME" metadata.annotations.externalClusterScript "$(base64 <$CEPH_EXTERNAL_SCRIPT_FILE)"
+
+    if [[ "${GENERATE_ROOK_CSV_FOR_TRACKING_ONLY}" != "true" ]]; then
+        # base64 has different behavior on linux vs macos
```
Comment (leseb) [inline cluster/olm/ceph/generate-rook-csv.sh:132]: what's the difference?

File: images/ceph/Makefile (hunk 5 above)
Comment (leseb) [reply images/ceph/Makefile:95]: How is it useful in debuging? Is it printing the commands?

**Review by leseb — COMMENTED** [review]
Why do we have to track all the RBACs too? This looks like a lot of duplication from `common.yaml`, isn't it?

**Comment by BlaineEXE** [comment]
@leseb 

I was wanting to revisit this for RBAC generation from the Helm chart (which will/could affect CSV generation). I think it may be best to track the CSV's RBAC for now, then start generating CSV RBAC using a Python script like https://github.com/rook/rook/blob/master/build/rbac/keep-rbac-yaml.py. 

If your previous comment (https://github.com/rook/rook/pull/8862#issuecomment-938628205) is true, then we should be including all the RBAC in the CSV, and we can stop using `# OLM: {BEGIN,END}` comments for the generation. This will show us what RBAC is being added compared to the existing set, and we will have a diff to check and see if anything needs to be pruned.

After that, we can decide if we should stop versioning the CSV RBAC or if we should keep it to verify changes to generated RBAC.

Unless you have a better idea for how we can keep track of what changes with generated CSV stuff?

**Comment by leseb** [comment]
> @leseb
> 
> I was wanting to revisit this for RBAC generation from the Helm chart (which will/could affect CSV generation). I think it may be best to track the CSV's RBAC for now, then start generating CSV RBAC using a Python script like https://github.com/rook/rook/blob/master/build/rbac/keep-rbac-yaml.py.
> 
> If your previous comment ([#8862 (comment)](https://github.com/rook/rook/pull/8862#issuecomment-938628205)) is true, then we should be including all the RBAC in the CSV, and we can stop using `# OLM: {BEGIN,END}` comments for the generation. This will show us what RBAC is being added compared to the existing set, and we will have a diff to check and see if anything needs to be pruned.
> 
> After that, we can decide if we should stop versioning the CSV RBAC or if we should keep it to verify changes to generated RBAC.
> 
> Unless you have a better idea for how we can keep track of what changes with generated CSV stuff?

No, I think tracking the changes as part of git is good.

**Comment by BlaineEXE** [comment]
Closing in favor of https://github.com/rook/rook/pull/9200

**Pull request closed without merging**